File: src/ship.lua
local Actor = require('engine.actor')
local Weapon = require('engine.weapon.weapon')
local AnimatedSprite = require('engine.sprite.animated-sprite')

---@class Ship:Actor
---@field weapon Weapon
Ship = setmetatable({}, { __index = Actor })
Ship.__index = Ship

---@param spriteEngine SpriteEngine
---@param x number
---@param y number
function Ship:new(spriteEngine, x, y)
    local image = love.graphics.newImage('assets/ship.png')
    local obj = Actor:new(spriteEngine, x, y, 200, AnimatedSprite:new(image, 80, 64, 0.1))
    obj.weapon = Weapon:new(spriteEngine, 0, 0, 1, obj)
    setmetatable(obj, Ship)
    return obj
end

---@param self Ship
---@param dt number
function Ship:update(dt)
    Actor.update(self, dt) -- Call the parent class's update method if needed
    self.weapon:update(dt)

    if love.keyboard.isDown('left') then
        self.x = self.x - self.speed * dt
    elseif love.keyboard.isDown('right') then
        self.x = self.x + self.speed * dt
    end

    if love.keyboard.isDown('up') then
        self.y = self.y - self.speed * dt
    elseif love.keyboard.isDown('down') then
        self.y = self.y + self.speed * dt
    end

    if love.keyboard.isDown('space') then
        self.weapon:fire()
    end
end

function Ship:draw()
    Actor.draw(self)
    self.weapon:draw()
end

return Ship

File: src/ship.ts
import { Image } from "love.graphics";
import { Actor } from "./engine/actor";
import { Engine } from "./engine/engine";
import { Input } from "./engine/input/input";
import { SpriteEngine } from "./engine/sprite-engine";
import { AnimatedSprite } from "./engine/sprite/animated-sprite";
import { Vector2 } from "./engine/tools";
import { Weapon } from "./engine/weapon/weapon";

let image: Image;
Engine.preload(() => {
  image = love.graphics.newImage("/assets/ship.png")
})

export class Ship extends Actor {
  weapon: Weapon;

  constructor(spriteEngine: SpriteEngine, pos: Vector2) {
    const animatedSprite = new AnimatedSprite(image, 40, 32, 0.1);
    super(spriteEngine, pos, 200, animatedSprite);

    this.weapon = new Weapon(spriteEngine, new Vector2(30, 10), 0.15, this);
  }

  private getDir(): { x: number; y: number } {
    let x = 0;
    let y = 0;

    for (const joystick of Input.getJoysticks()) {
      const axisX = joystick.getGamepadAxis("leftx");
      const axisY = joystick.getGamepadAxis("lefty");

      if (Math.abs(axisX) > 0.2) {
        x = axisX;
      }

      if (Math.abs(axisY) > 0.2) {
        y = axisY;
      }
      if (joystick.isGamepadDown("start")) {
        love.event.quit(0);
      }
    }
    return { x, y };
  }

  private isFiring(): boolean {
    for (const joystick of Input.getJoysticks()) {
      if (joystick.isGamepadDown("b")) {
        return true;
      }
    }
  }

  update(dt: number) {
    super.update(dt); // Call the parent class's update method
    this.weapon.update(dt);

    const dir = this.getDir();

    this.pos.x += dir.x * this.speed * dt;
    this.pos.y += dir.y * this.speed * dt;

    if (this.isFiring()) {
      this.weapon.fire();
    }
  }

  draw() {
    super.draw(); // Call the parent class's draw method
    this.weapon.draw();
  }
}
